waiter/src/waiter/state.clj

```diff
@@ -1303,3 +1308,114 @@
+                                                                :else true))))]
+                 (->InstanceTracker service-id known-instance-ids' scheduling-instance-timer-contexts' starting-instance-trackers'))))))
+
+(defn start-instance-startup-stats-maintainer
+  "go block to collect metrics on the schedule/startup overhead of waiter services.
+
+  Updated state for the router is read from `router-state-chan`."
+  [router-state-chan service-id->service-description-fn]
+  (let [exit-chan (async/chan 1)
+        query-chan (au/latest-chan)
+        update-state-timer (metrics/waiter-timer "state" "instance-startup-stats-maintainer" "update-state")]
+    (async/go
+      (try
+        (loop [{:keys [known-service-ids service-instance-trackers] :as current-state}
+               {:known-service-ids #{}
+                :service-instance-trackers nil}]
+          (let [new-state
+                (async/alt!
+                  exit-chan
+                  ([message]
+                   (if (= :exit message)
+                     (do
+                       (log/warn "stopping instance-startup-stats-maintainer")
+                       (comment "Return nil to exit the loop"))
+                     current-state))
+
+                  router-state-chan
+                  ([router-state]
+                   (timers/start-stop-time!
+                     update-state-timer
+                     (let [{:keys [service-id->healthy-instances service-id->my-instance->slots service-id->unhealthy-instances]} router-state
+                           incoming-service-ids (set (keys service-id->my-instance->slots))
+                           new-service-ids (set/difference incoming-service-ids known-service-ids)
+                           removed-service-ids (set/difference known-service-ids incoming-service-ids)
+                           service-instance-trackers' (update-instance-trackers
+                                                        service-instance-trackers new-service-ids removed-service-ids
+                                                        service-id->healthy-instances service-id->unhealthy-instances
+                                                        service-id->service-description-fn)]
+                       {:known-service-ids incoming-service-ids
+                        :service-instance-trackers service-instance-trackers'})))
+
+                  query-chan
+                  ([{:keys [cid response-chan]}]
+                   (cid/cinfo cid (str "returning current instance-startup state " current-state))
+                   (async/put! response-chan current-state)
+                   current-state)
+                  :priority true)]
+            (when new-state
+              (recur new-state))))
+        (catch Exception e
+          (log/error e "Fatal error in instance-startup-stats-maintainer")
+          (System/exit 1))))
```

Comment:
Is it possible to keep running without exiting?

Reply:
Fixed.